### ISSUE TITLE
Feature request - Make streamfield blocks collapsible in admin

### ISSUE THREAD

**Issue by mattots** [wagtail/wagtail#1336]
It would be nice to be able to toggle the content of StreamField blocks in the admin - so with all blocks collapsed you'd just see a list of the block type labels (something like the attached), which, when clicked on, would expand to show the corresponding fields. This would be useful in situations where there are a lot of blocks on a page and you don't want to have to scroll endlessly to find the fields you want to edit.

![screenshot](https://cloud.githubusercontent.com/assets/301729/7774311/322a638e-00a2-11e5-923e-65726189ca90.png)

![screenshot2](https://cloud.githubusercontent.com/assets/301729/7774372/d5009916-00a2-11e5-80ce-b7aa01402b67.png)

**Comment by gasman** [comment]
Hi @mattots,

This seems a bit counter-productive to me... surely if there's a particular image or bit of text you want to edit, you're better off scrolling through the interface looking for the image/text in question, rather than having to think about what type of block it is ("wait, is this an image, or a featured image...?") and clicking into them to find it?

**Comment by mattots** [comment]
A lot of the sites we built (and still maintain) before we started using Wagtail had something similar to (but rather more clunky than) the StreamField with collapsible blocks, so we've got quite a bit of experience of how clients use this sort of functionality, and it seems to be something that they find useful. I don't think it should necessarily be enforced, but an option to make the blocks collapsible via a parameter, and possibly to define whether they start expanded or collapsed, would be nice.

**Comment by kidlj** [comment]
when I use StreamField I feel eager to own this feature, truly.

**Comment by Ernest0x** [comment]
This feature is surely very useful, especially when there is a great number of blocks in the streamfield. @gasman is right in that seeing only the block's type is not enough to find the exact block you want to edit. But this can be easily improved by optionally using the value of a text field as a title while in collapsed state. So, for example, image blocks in the example above could be configured to have a title consisting of the 'caption' field  value. What do you thing?

**Comment by gasman** [comment]
@Ernest0x That sounds like a good idea, although it would add quite a bit of complexity to the implementation - you'd need some way to specify which field to use as the title, as Wagtail doesn't know which fields are the important ones.

To be honest, I'm not massively keen on this proposal, as it feels like "sweeping the real problem under the carpet". If you need to hide fields in order to make the edit interface manageable, then I'd take that as a warning sign that your information architecture could be improved. (As a rough rule of thumb - the vertical space that your edit interface takes up should not really be substantially more than the height of the page as seen on the site's front-end.)

Maybe that improvement is something that you can do as a site implementer, such as moving some data into sub-pages or snippets... maybe it calls for a new feature in Wagtail that hasn't been identified yet. Quite possibly, it's a bit of both. I feel that collapsible blocks are a similar situation to rich text fields - a convenient hack that solves the immediate problem, but closes your mind to the possibility of better ways of organising your data.

In the case of @mattots's example, I would point the finger at the 'class' and 'ID' fields - those are low-level HTML details that, ideally, editors wouldn't be dealing with. (That's not a criticism... as I say, it may well point to an as-yet-unidentified missing feature of Wagtail.) If you have several different types of image that are distinguished by different CSS/JS behaviour, then perhaps it would be more natural to make those into distinct block types - like, say, 'photo' and 'infographic' - instead of a single 'image' type.

Alternatively, one design concept we had floating around some months ago was to have a 'cog' icon in the corner of the block, to give access to the lesser-used configuration options (like I suspect 'class' and 'ID' are here) so that they aren't taking up space in the main body of the page. Either way - if we can find a way to eliminate those fields, we'd immediately halve the height of that block.

I know it probably sounds like I'm replacing a simple solution with a more complicated one, but I suspect that collapsible blocks would be a deceptively complex thing to implement (due to the need to introduce new configuration options, and possibly remember the open/closed state of each panel on a per-user basis), and my gut feeling is that there are simpler solutions that will address the root cause of the problem more effectively.

**Comment by mattots** [comment]
I really like the idea of the cog. You're right in pointing out the 'class' and 'id' fields as being something that most editors wouldn't/shouldn't need to worry about and in our situation its fairly rare that they actually are required (although having them there has proved useful at times which is why we keep them in by default). But being able to hide away lesser used options is a nice idea that could help keep things simpler and reduce page length.

**Comment by Ernest0x** [comment]
@gasman I would suggest the use of a Meta class attribute named "admin_title_fieldname" or something like that. Remembering of open/closed state for panels is not that useful either. All panels would be loaded in a collapsed state by default, unless there are validation errors. I do not think this would add that much complexity to neither the implementation or the admin definition process.

Also, there may be great difference between data definition (admin), which is always vertical, and data representation (frontend) which may be horizontal. So, the rule of thumb you are mentioning is not practical in many cases.

Furthermore, the number of fields in a block may not be relevant. Usually, the problem appears when there is repeatability of same blocks. In the example above, the Image block, even if it only consisted of 'image', 'caption' and 'link url' fields, it would still take pretty much vertical space if it had to be repeated multiple times.

Another suggestion I am thinking now is the introduction of "container" panels. A "container" panel would act as a way to group some panels together and give them some UI possibilities like collapsing, moving up and down as a whole, removing, etc.

**Comment by jaroel** [comment]
This would probably make managing larger lists easier. Refs #2264.

**Comment by alexgleason** [comment]
@heymonkeyriot has been secretly working on some designs that could improve giant streamfield blocks

**Comment by SalahAdDin** [comment]
:+1:

**Comment by heymonkeyriot** [comment]
Though what I've been working on moves in a slightly different direction to the suggestion that @mattots made above (I hadn't seen this issue before) I think making them collapsible could make a lot of sense (and is sort of in keeping with how `inline blocks` can be open or closed). But collapsing them also feels like it moves away from my understanding of the original goal of streamfield, which was to have the content presented in a way that offered the least possible distraction to the editor.  I showed  @alexgleason some very rough mockups because I could be there to explain them, will share them virtually when they're a bit more together :)

**Comment by Eraldo** [comment]
How about added an optional "collapsible" class to streamfieldpanels and adding an "expand all" button.
As I see it that would solve the preference of collapsing without losing the search/overview capability.
What do you think?

**Comment by BertrandBordage** [comment]
As part of the Wagtail’s first Kickstarter campaign, I’m working on this!
https://www.kickstarter.com/projects/noripyt/wagtails-first-hatch

**Comment by commadelimited** [comment]
I'm interested in this feature as well.

Additionally it would be great if you could collapse top level sections of the site as well. For example the downward facing arrow to the left of the red section **already** makes me think I should be able to click to collapse.

**Comment by zerolab** [comment]
Hi @commadelimited,

I am not mistaken this can be done with https://github.com/wagtail/wagtail-react-streamfield#meta-attributes-or-passed-to-__init__ does this, so when it makes it in Wagtail, the functionality will be there.

**Comment by commadelimited** [comment]
@zerolab oh wow, nice! Thank you for the mention.

**Comment by jkevingutierrez** [comment]
+1

**Comment by gasman** [comment]
Implemented as part of the telepath StreamField rewrite in 2.13.

**Comment by bitwise-gmbh** [comment]
> Implemented as part of the telepath StreamField rewrite in 2.13.

Could you explain, how StreamField block can now be collapsed since the implementation in 2.13?

**Comment by gasman** [comment]
@bitwise-gmbh That isn't implemented yet - there's a separate ticket for that at #7705.
